WIP: Starknet Snap Rough Skeleton (#1509)
MetaMask/metamask-docs: wallet/how-to/use-non-evm-networks/starknet/send-starknet-transactions.md

[Connect to Starknet](connect-to-starknet.md) from your dapp.

## Send a transaction

Send a transaction using the [`starknet.account.execute()`] (https://starknetjs.com/docs/api/classes/accountinterface/#execute) function:

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript
  const sendStarknetTransaction = async (wallet, contractAddress, entrypoint, calldata) => {
    try {
      if(wallet?.isConnected !== true){
        throw("Wallet not connected");
      } 
    
      // Send the transaction.
      const result = await wallet?.account?.execute({
        contractAddress: contractAddress,  // The address of the contract.
        entrypoint: entrypoint,            // The function to call in the contract.
        calldata: calldata                 // The parameters to pass to the function.
      });
      console.log("Transaction successful:", result);
      return result;
    } catch (error) {
      console.error("Error sending transaction:", error);
    }
  };
  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript
  const sendStarknetTransaction = async (contractAddress, contractFuncName, contractCallData, address, chainId, maxFee = null) => {

>     if (typeof window.ethereum === "undefined" || !window.ethereum.isMetaMask) {
>       throw new Error("MetaMask not detected or Snaps not supported");
>     }
> 
>     try {
>       const response = await window.ethereum.request({

Reviewer: ```suggestion
    const provider = await getEip6963Provider();
    
    if (!provider) {
      throw new Error("MetaMask not detected or Snaps not supported");
    }

    try {
      const response = await provider.request({
```

add eip6963 handle

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript
  import { connect } from "get-starknet";

  const connectStarknetAccount = async () => {
    const starknet = await connect();
    await starknet.enable();  // Prompts the user to connect their Starknet account using MetaMask
    return starknet;
  };

  const sendStarknetTransaction = async (contractAddress, entrypoint, calldata) => {
    try {
      const starknet = await connectStarknetAccount();  // Ensure the account is connected

      // Send the transaction
      const result = await starknet.account.execute({
        contractAddress: contractAddress, 
        entrypoint: entrypoint,            
        calldata: calldata                 
      });

      console.log("Transaction successful:", result);
      return result;
    } catch (error) {
      console.error("Error sending transaction:", error);
    }
  };

  const contractAddress = "0xYourContractAddress";  
  const entrypoint = "your_function_name";          
  const calldata = [/* your function arguments */]; 

  sendStarknetTransaction(contractAddress, entrypoint, calldata);
  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript

>   const connectStarknetAccount = async () => {

Reviewer: ```suggestion
  const connectStarknetAccount = async (provider) => {
```

>     if (typeof window.ethereum === "undefined" || !window.ethereum.isMetaMask) {
>       throw new Error("MetaMask not detected or Snaps not supported");
>     }
> 
>     try {
>       await window.ethereum.request({

Reviewer: ```suggestion
    try {
      await provider.request({
```

        method: "wallet_requestSnaps",
        params: {
          "npm:@consensys/starknet-snap": {}
        }
      });
      console.log("Starknet Snap connected");
    } catch (error) {
      console.error("Error connecting to Starknet Snap:", error);
      throw error;
    }
  };

  const sendStarknetTransaction = async (contractAddress, contractFuncName, contractCallData, address, chainId, maxFee = null) => {
    try {

>       await connectStarknetAccount();

Reviewer: ```suggestion
      const provider = await getEip6963Provider();
      
      if (!provider) {
        throw new Error("MetaMask not detected or Snaps not supported");
      }
      
      await connectStarknetAccount(provider);
```

      
      const requestParams = {
        address,
        calls: {
            contractAddress,
            entrypoint,
            calldata,
        }
        chainId
      };

      if (maxFee) {
        requestParams.details = {
            maxFee
        }; // Include maxFee only if it's provided.
      }

>       const response = await window.ethereum.request({

Reviewer: ```suggestion
      const response = await provider.request({
```

        method: "wallet_invokeSnap",
        params: {
          snapId: "npm:@consensys/starknet-snap",
          request: {
            method: "starkNet_executeTxn"
            params: requestParams
          }
        }
      });

      console.log("Transaction sent:", response);
      return response;
    } catch (error) {
      console.error("Error sending transaction:", error);
      throw error;
    }
  };

  // Example usage
  const contractAddress = "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4";
  const contractFuncName = "transfer";
  const contractCallData = ["0x5B38Da6a701c568545dCfcB03FcB875f56beddC4", "1000"];
  const senderAddress = "0xb60e8dd61c5d32be8058bb8eb970870f07233155";
  const maxFee = "1000000000000000"; // Optional

> const chainId = '0x534e5f5345504f4c4941';

Reviewer: ```suggestion
  const chainId = '0x534e5f5345504f4c4941';
```

## Send a transaction

Send a transaction using the
[`starknet.account.execute()`](https://starknetjs.com/docs/api/classes/accountinterface/#execute)
function (with `get-starknet`) or the
[`starknet_executeTxn`](../../../reference/non-evm-apis/starknet-snap-api.md#starknet_executetxn)
method (with `wallet_invokeSnap`):

<Tabs>
  <TabItem value="get-starknet" default>

  ```javascript
  const sendStarknetTransaction = async (wallet, contractAddress, entrypoint, calldata) => {
    try {
      if(wallet?.isConnected !== true){
        throw("Wallet not connected");
      } 
    
      // Send the transaction.
      const result = await wallet?.account?.execute({
        contractAddress: contractAddress,  // The address of the contract.
        entrypoint: entrypoint,            // The function to call in the contract.
        calldata: calldata                 // The parameters to pass to the function.
      });
      console.log("Transaction successful:", result);
      return result;
    } catch (error) {
      console.error("Error sending transaction:", error);
    }
  };
  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript

>   const sendStarknetTransaction = async (contractAddress, contractFuncName, contractCallData, address, chainId, maxFee = null) => {

Reviewer 2: Should these be updated to match the variables used later in this code snippet for the `executeTxn` params?

In general this code sample should be reviewed for syntax and accuracy.

    }
  };

  const contractAddress = "0xYourContractAddress";  
  const entrypoint = "your_function_name";          
  const calldata = [/* your function arguments */]; 

  sendStarknetTransaction(contractAddress, entrypoint, calldata);
  ```

  </TabItem>
  <TabItem value="wallet_invokeSnap">

  ```javascript
  const connectStarknetAccount = async (provider) => {
    try {
      await provider.request({
        method: "wallet_requestSnaps",
        params: {
          "npm:@consensys/starknet-snap": {}
        }
      });
      console.log("Starknet Snap connected");
    } catch (error) {
      console.error("Error connecting to Starknet Snap:", error);
      throw error;
    }
  };

   const sendStarknetTransaction = async (contractAddress, contractFuncName, contractCallData, address, chainId, maxFee = null) => {
    try {
        const provider = await getEip6963Provider();
      
      if (!provider) {
        throw new Error("MetaMask not detected or Snaps not supported");
      }
      
      await connectStarknetAccount(provider);

      const requestParams = {
        contractAddress,
        contractFuncName,
        contractCallData,

>         senderAddress

Reviewer 2: Should these be updated to match the `executeTxn` params?

      };

      if (maxFee) {
        requestParams.details = {
            maxFee
        }; // Include maxFee only if it's provided.

      const response = await getEip6963Provider.request({
        method: "wallet_invokeSnap",
        params: {
          snapId: "npm:@consensys/starknet-snap",
          request: {
            method: "starkNet_executeTxn"
            params: requestParams
          }
        }
      });

      console.log("Transaction sent:", response);
      return response;
    } catch (error) {
      console.error("Error sending transaction:", error);
      throw error;
    }
  };

  // Example usage
  const contractAddress = "0x5B38Da6a701c568545dCfcB03FcB875f56beddC4";
  const contractFuncName = "transfer";
  const contractCallData = ["0x5B38Da6a701c568545dCfcB03FcB875f56beddC4", "1000"];
  const senderAddress = "0xb60e8dd61c5d32be8058bb8eb970870f07233155";
  const maxFee = "1000000000000000"; // Optional
  const chainId = "0x534e5f5345504f4c4941";

>   sendStarknetTransaction(contractAddress, contractFuncName, contractCallData, senderAddress, chainId, maxFee)

Reviewer 2: Similarly, these might need to be updated to match the listed `executeTxn` params.

---
description: Send Starknet transactions in MetaMask.
sidebar_position: 4
---

# Send Starknet transactions

> After the account is connected, you can send a transaction using the `starknet.invoke()` function:

Reviewer 3: The `connect-to-starknet` section shows how to get access to the wallet object that we use here 

```suggestion
After the account is connected, you can send a transaction using the `wallet.account.execute` function:
```

```javascript

> const sendStarknetTransaction = async (contractAddress, entrypoint, calldata) => {
>   try {
>     const starknet = await connectStarknetAccount();  // Ensure the account is connected
> 
>     // Send the transaction
>     const result = await starknet.invoke({
>       contractAddress: contractAddress,  // The address of the contract
>       entrypoint: entrypoint,            // The function to call in the contract
>       calldata: calldata                 // The parameters to pass to the function
>     });
> 
>     console.log('Transaction successful:', result);
>     return result;
>   } catch (error) {
>     console.error('Error sending transaction:', error);
>   }
> };
> ```

Reviewer 3: ```suggestion
const sendStarknetTransaction = async (wallet, contractAddress, entrypoint, calldata) => {
  try {
    if(wallet?.isConnected !== true){
      throw('Wallet not connected');
    } 
  
    // Send the transaction
    const result = await wallet?.account?.execute({
      contractAddress: contractAddress,  // The address of the contract
      entrypoint: entrypoint,            // The function to call in the contract
      calldata: calldata                 // The parameters to pass to the function
    });

    console.log('Transaction successful:', result);
    return result;
  } catch (error) {
    console.error('Error sending transaction:', error);
  }
};
```
```

## Simplified Example

```javascript

> import { getStarknet } from 'get-starknet';

Reviewer 3: ```suggestion
import { connect } from 'get-starknet';
```

const connectStarknetAccount = async () => {

>   const starknet = getStarknet();

Reviewer 3: ```suggestion
  const starknet = await connect();
```

  await starknet.enable();  // Prompts the user to connect their Starknet account via MetaMask
  return starknet;
};

const sendStarknetTransaction = async (contractAddress, entrypoint, calldata) => {
  try {
    const starknet = await connectStarknetAccount();  // Ensure the account is connected

    // Send the transaction

>     const result = await starknet.invoke({

Reviewer 3: ```suggestion
    const result = await starknet.account.execute({
```

  <TabItem value="wallet_invokeSnap">

  ```javascript

>   const sendStarknetTransaction = async (address, chainId, contractAddress, entrypoint, calldata, maxFee = null) => {

Reviewer 2: Updated these params
